src/public.ts

```diff
@@ -0,0 +1,255 @@
+import { StreamDBConsumerClient } from "./stream-db/streamDbDatabase";
+import { Timestamp } from "./model/timestamp";
+import { execAndReturnWithRetry } from "./utils";
+import axios from "axios";
+import {strict as assert} from "assert";
+import { PausableStream } from "./stream-db/pausableStream";
+
+
+export class StreamClient {
+    private readonly registry: StreamRegistry;
+    private clientCache: Record<string, StreamDBConsumerClient>
+  
+    public constructor(
+      private readonly options: StreamClientOptions,
+    ) {
+      this.registry = options.registry ?? new RemoteStreamRegistry();
+      this.clientCache = {}
+    }
+  
+    public async findOffset(stream: string, params: {height: number, timestamp?: Timestamp}): Promise<Offset | undefined> {
+      // 1. get all endpoints having the stream
+      // 2. try every to find offset
+      // 3. return first found offset
+ 
+      const endpoints = await this.registry.findStreamEndpoints(stream, new Offset("", BigInt(params.height), params.timestamp ?? Timestamp.zero))
+      for (const endpoint of endpoints) {
+        const client = this.getStreamConsumerClient(endpoint.uri)
+        const offset = await client.findOffset(stream, BigInt(params.height || 0), params.timestamp)
+        if (offset !== undefined) {
+            return offset
+        }
+        return undefined
+      }
+    }
+  
+    //return stream matching the name 
+    public async getStream(name: string): Promise<Stream> {
+      // 1. return first endpoint having the stream
+      // 2. throw error if stream not found
+      try {
+        return await this.registry.getStream(name)
+      } catch(e) {
+          throw new Error("Stream not found" + e)
+      }
+    }
+  
+    public async fetchEvents(streamName: string, offset: Offset, count: number, direction: "next" | "last"): Promise<StreamEvent[]> {
+      // 1. find endpoint having the offset
+      // 2. proxy request to it
+      // 3. note, it's worth keeping clients to every streamdb endpoint cached (as well as grpc connections to them)
+      const endpoints: StreamEndpoint[] = await this.registry.findStreamEndpoints(streamName, offset)
```

Comment:
So, right now every client call makes a request to stream registry. And as far as I understand, the results of stream discovery are pretty much immutable until a new chunk of a stream is produced. So maybe we should cache the result of a discovery and only do it once for a stream? I think it would reduce network times significantly because two requests to different locations sound like a long latency.

Reply:
Results aren't immutable, but added a simple cache that ttimes out values so that requests do not need to be duplicated for same stream requests